File: MiniDOMOffice/Magazzino/Articolo/AttributiArticolo/Collection/AttributiArticoloCollection.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using minidom;
using minidom.repositories;
using static minidom.Sistema;
using static minidom.Anagrafica;
using static minidom.Store;


namespace minidom
{
    public partial class Store
    {

        /// <summary>
        /// Collezione degli attributi specificato per una istanza di un Articolo
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class AttributiArticoloCollection 
            : CKeyCollection<AttributoArticolo>
        {
            [NonSerialized]
            private Articolo m_Articolo;

            /// <summary>
            /// Costruttore
            /// </summary>
            public AttributiArticoloCollection()
            {
                m_Articolo = null;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="Articolo"></param>
            public AttributiArticoloCollection(Articolo Articolo) : this()
            {
                Load(Articolo);
            }

            /// <summary>
            /// Restituisce l'articolo a cui appartiene la collezione
            /// </summary>
            public Articolo Articolo
            {
                get
                {
                    return m_Articolo;
                }
            }

            /// <summary>
            /// Imposta l'articolo
            /// </summary>
            /// <param name="value"></param>
            protected internal virtual void SetArticolo(Articolo value)
            {
                m_Articolo = value;
                if (value is object)
                {
                    foreach (AttributoArticolo item in this)
                        item.SetArticolo(value);
                }
            }

            /// <summary>
            /// OnInsert
            /// </summary>
            /// <param name="index"></param>
            /// <param name="value"></param>
            protected override void OnInsert(int index, AttributoArticolo value)
            {
                if (m_Articolo is object)
                    value.SetArticolo(m_Articolo);
                base.OnInsert(index, value);
            }

            /// <summary>
            /// OnSet
            /// </summary>
            /// <param name="index"></param>
            /// <param name="oldValue"></param>
            /// <param name="newValue"></param>
            protected override void OnSet(int index, AttributoArticolo oldValue, AttributoArticolo newValue)
            {
                if (m_Articolo is object)
                    newValue.SetArticolo(m_Articolo);
                base.OnSet(index, oldValue, newValue);
            }

            /// <summary>
            /// Carica
            /// </summary>
            /// <param name="articolo"></param>
            protected internal void Load(Articolo articolo)
            {
                if (articolo is null)
                    throw new ArgumentNullException("articolo");
               
                this.Clear();

                this.m_Articolo = articolo;

                if (DBUtils.GetID(articolo, 0) == 0)
                    return;

                using (var cursor = new AttributoArticoloCursor())
                {
                    cursor.IgnoreRights = true;
                    cursor.IDArticolo.Value = DBUtils.GetID(articolo, 0);
                    cursor.Stato.Value = ObjectStatus.OBJECT_VALID;
                    while (cursor.Read())
                    {
                        Add(cursor.Item.NomeAttributo, cursor.Item);
                    }
                }

                this.Sort();
            }
        }
    }
}

File: MiniDOMOffice/Magazzino/Articolo/AttributiArticolo/AttributiArticoloCollection.vb
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class Office
     
    ''' <summary>
    ''' Collezione degli attributi specificato per una istanza di un Articolo
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class AttributiArticoloCollection
        Inherits CKeyCollection(Of AttributoArticolo)

        <NonSerialized> Private m_Articolo As Articolo

        Public Sub New()
            Me.m_Articolo = Nothing
        End Sub

        Public Sub New(ByVal Articolo As Articolo)
            Me.New()
            Me.Load(Articolo)
        End Sub

        Public ReadOnly Property Articolo As Articolo
            Get
                Return Me.m_Articolo
            End Get
        End Property

        Protected Friend Overridable Sub SetArticolo(ByVal value As Articolo)
            Me.m_Articolo = value
            If (value IsNot Nothing) Then
                For Each item As AttributoArticolo In Me
                    item.SetArticolo(value)
                Next
            End If
        End Sub

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Articolo IsNot Nothing) Then DirectCast(value, AttributoArticolo).SetArticolo(Me.m_Articolo)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Articolo IsNot Nothing) Then DirectCast(newValue, AttributoArticolo).SetArticolo(Me.m_Articolo)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Public Sub Load(ByVal articolo As Articolo)
            If (articolo Is Nothing) Then Throw New ArgumentNullException("articolo")
            Me.Clear()
            Me.m_Articolo = articolo
            If (GetID(articolo) = 0) Then Exit Sub
            Using cursor As New AttributoArticoloCursor()
                cursor.IgnoreRights = True
                cursor.IDArticolo.Value = GetID(articolo)
                cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                While Not cursor.EOF
                    Me.Add(cursor.Item.NomeAttributo, cursor.Item)
                    cursor.MoveNext()
                End While
            End Using
            Me.Sort()
        End Sub


    End Class

End Class


